Heimdallr/OAuthAuthorizationCodeHandler.swift

```diff
@@ -0,0 +1,142 @@
+import Foundation
+import UIKit
+import SafariServices
+import Result
+
+/// The request could not be authorized (e.g., no refresh token available).
+public let HeimdallrErrorAuthorizationCanceled = 101
+
+private enum CompletionType {
+    case authorizationCode(AuthorizationCodeCompletion)
+    case accessToken(AccessTokenCompletion)
+    case none
+}
+
+extension CompletionType {
+
+    func oAuthAccessToken(fromURL url: URL) -> OAuthAccessToken? {
+
+        guard let paramters = url.fragmentParameters,
+            let token = paramters["access_token"] else { return nil }
+
+        return OAuthAccessToken(accessToken: token)
+    }
+
+    func handle(url: URL) {
+        switch self {
+        case let .authorizationCode(completion):
+            guard let authorizationCode = url.queryParameters?["code"] else {
+                completion(.failure(NSError(domain: "", code: 0, userInfo: nil)))
+                return
+            }
+            completion(.success(authorizationCode))
+        case let .accessToken(completion):
+            guard let oAuthAccessToken = oAuthAccessToken(fromURL: url) else {
+                completion(.failure(NSError(domain: "", code: 0, userInfo: nil)))
+                return
+            }
+            completion(.success(oAuthAccessToken))
+        case .none:
+            break
+        }
+    }
+
+    func fail(error: NSError) {
```

Comment:
Maybe both methods should jus be named `complete(with:)` accepting different types? Might be a better fit for a `CompletionType` to `complete` rather then `fail` or `handle`. That being said: Probably there is a nicer way of abstracting this all together. Maybe we can pair on this at some point?